File: 437.路径总和-iii.py
#
# @lc app=leetcode.cn id=437 lang=python
#
# [437] 路径总和 III
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        if not root :
            return 0
        
        stack = [(root, [root.val])]
        res = 0

        while stack:
            node, temp = stack.pop()
            res += temp.count(sum)
            temp += [0]
            if node.left:
                arr = [i+node.left.val for i in temp]
                stack.append((node.left,arr))
            if node.right:
                arr = [i+node.right.val for i in temp]
                stack.append((node.right, arr))
        return res
        
# @lc code=end

File: 437.路径总和-iii.cpp
/*
 * @lc app=leetcode.cn id=437 lang=cpp
 *
 * [437] 路径总和 III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    void dfs(TreeNode* node, int sum){
        if (node == nullptr) return ;
        sum -= node->val;
        if(sum == 0)
            ans++;
        dfs(node->left, sum);
        dfs(node->right, sum);

    }
    int pathSum(TreeNode* root, int sum) {
        if(root == nullptr)
            return ans;
        dfs(root, sum);
        pathSum(root->left, sum);
        pathSum(root->right, sum);
        return ans;

    }
};
// @lc code=end

